### DISCUSSION TITLE
Parallel build with MSVC?

### DISCUSSION THREAD

**Discussion by asmaloney** [community/community#26485] [Actions]
<p>I am using <em>windows-latest</em> with the <em>Visual Studio 16 2019</em> generator and cmake.</p>
<p>I’m confiuring with cmake like this:</p>
<pre><code>cmake -B build -G "${{ matrix.config.generator }}" ...
</code></pre>
<p>and then building:</p>
<pre><code>cmake --build build --parallel
</code></pre>
<p>But it doesn’t seem to be using multiple cores… (at least the builds take about twice as long as macOS)</p>
<p>How do I tell it to use all the available cores?</p>
<p>On the macOS and Linux side I use ninja, so it magically works.</p>
<p>(Disclaimer: I’m not a Windows guy :slight_smile: )</p>

**Comment by Yanjingzhu** [comment]
<p>I am sorry for the delay response. To be honest I am not a cmake expert . Would you mind asking for help from MSVC team for the right configuration to make cmake build run in parallel on windows OS? </p>

**Comment by asmaloney** [comment]
<p>Thank you.</p>
<p>I figured this would be an extremely common use-case/request and that it would require knowledge of GitHub Actions to get set up properly (how to get the number of cores and that sort of thing) which is why I’m asking here.</p>
<p>I have no idea where/how to deal with anything MSVC. I’m not about to try to navigate their bureaucracy - I’ve wasted enough of my life dealing with Apple lately :slight_smile:</p>

**Comment by asmaloney** [comment, marked as answer]
<p>OK I have something working: switch to the <em>ninja</em> generator :slight_smile:</p>
<p>After installing ninja, I needed to do the following:</p>
<pre><code>matrix:
        config:
          - {
              name: "Windows MSVC",
              os: windows-latest,
              generator: "Ninja",
           }
...
    steps:
      ...
      - name: Configure MSVC console (Windows)
        if: matrix.config.os == 'windows-latest'
        uses: ilammy/msvc-dev-cmd@v1

      - name: Set environment for MSVC (Windows)
        if: matrix.config.os == 'windows-latest'
        run: |
          # Set these env vars so cmake picks the correct compiler
          echo "::set-env name=CXX::cl.exe"
          echo "::set-env name=CC::cl.exe"
...
</code></pre>
<p>Then calling:</p>
<pre><code>cmake -G "${{ matrix.config.generator }}" ...
</code></pre>
<p>will use ninja on Windows.</p>
<p>Didn’t speed things up by too much for me, but it provides nicer output!</p>
